### ISSUE TITLE
[API Proposal]: Custom the dotnet runtime root folder path

### ISSUE THREAD

**Issue by lindexi** [dotnet/runtime#64430]
### Background and motivation

My team will publishing some applications using our special version of the dotnet runtime. We want to share the special dotnet runtime version files. 

But as we all know, the default `AppHost.exe` can only replace the MainDll path with `c3ab8ff13720e8ad9047dd39466b3c89 74e592c2fa383d4a3960714caef0c4f2` string, and the `dotnet runtime` search path can not be custom.

Today, my friend [walterlv](https://github.com/walterlv/) add the `DOTNET_ROOT` placeholder to the AppHost in <https://github.com/dotnet-campus/dotnetCampus.AppHost> . See https://github.com/dotnet-campus/dotnetCampus.AppHost/blob/main/src/dotnetCampus.AppHost/Patches/v6.0.1/0001-Add-dotnet_root-placeholder.patch

And we can customize the dotnet runtime folder path by this feature.

```xml
// csproj

  <PropertyGroup>
    <!-- The special dotnet runtime folder path. -->
    <DCAppHostDotnetRoot>runtime</DCAppHostDotnetRoot>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="dotnetCampus.AppHost" Version="1.0.0-alpha04" />
  </ItemGroup>
```

Can this feature be merged into the `dotnet runtime`? 

### API Proposal

The SDK can read the property, such as `AppHostDotnetRoot`. It will write the value from the `AppHostDotnetRoot` property to the AppHost.exe file, and the AppHost.exe will use the special dotnet runtime root folder via the value of the `AppHostDotnetRoot` property.  It will do nothing when the developer do not set the `AppHostDotnetRoot` property.

### API Usage

We can tell the SDK where the DotNET Root folder is via a property in csproj .

```xml
  <PropertyGroup>
    <!-- The special dotnet runtime folder path. -->
    <AppHostDotnetRoot>runtime</AppHostDotnetRoot>
  </PropertyGroup>
```

**Comment by agocke** [comment]
@lindexi Very interesting. Is the idea that you're shipping multiple apps with a shared runtime and you want to use the apphost to find the runtime instead of setting DOTNET_ROOT?

Could you give an example of how you use this? How would your application be laid out on disk and deployed?

**Comment by lindexi** [comment]
@agocke

> Is the idea that you're shipping multiple apps with a shared runtime and you want to use the apphost to find the runtime instead of setting DOTNET_ROOT?

Yes, my teams will publish some applications. And we should not use the self-contained in all the application. The self-contained application will waste a lot of disk space.

> Could you give an example of how you use this?

 Just as the https://github.com/dotnet-campus/dotnetCampus.AppHost 

> How would your application be laid out on disk and deployed?

- C:\Program Files\CompanyName\DotNETRuntime\[Version]\
- C:\Program Files\CompanyName\Produce1\
- C:\Program Files\CompanyName\Produce2\
- C:\Program Files\CompanyName\Produce3\

The application file layout is:

- C:\Program Files\CompanyName\ProduceN\ApplicationHost.exe
- C:\Program Files\CompanyName\ProduceN\ApplicationEntry.dll
- C:\Program Files\CompanyName\ProduceN\1.0.0\Business.dll
- C:\Program Files\CompanyName\ProduceN\2.0.0\Business.dll
- C:\Program Files\CompanyName\ProduceN\2.0.1\Business.dll

The ApplicationHost.exe is a `AppHost.exe` file. The ApplicationHost.exe will find and load the dotnet runtime from `C:\Program Files\CompanyName\DotNETRuntime\`.

The ApplicationHost.exe will run the application Main method in `ApplicationEntry.dll`.

The ApplicationEntry.dll will select the special application business version to run by the C# code. 

And we can update the Application business code by download the new Business.dll to the new version folder.

**Comment by lindexi** [comment]
Why do we want to customize the dotnet runtime root folder? 

1. We plan to use our special version of the dotnet runtime.
2. We are afraid of polluting the default share dotnet runtime folder by other applications, such as some evil application that will replace the dotnet runtime with their advertisement.

**Comment by Symbai** [comment]
Sounds extremely similar to https://github.com/dotnet/sdk/issues/10366 and https://github.com/dotnet/runtime/issues/3453

**Comment by vitek-karas** [comment]
To me this sounds almost exactly like https://github.com/dotnet/runtime/issues/53834 - at least the part about sharing the runtime.

**Comment by Symbai** [comment]
@vitek-karas Allowing developers to specify runtime (sub) folder would probably solve all three at once. If not after adding this it probably only takes very small changes to solve the other requests as well. Seems there is a high demand on such feature, whether its because of sharing runtime or just getting rid of the 250 files mess.

**Comment by vitek-karas** [comment]
I think the interesting question is what should be the SDK behavior around this.
As proposed, there's no SDK interaction - which basically means the produced app won't run on its own (since the apphost will look for the runtime at a location where the SDK didn't put anything).
And then there's the framework dependent versus self-contained difference.

Personally I could see doing basically what the issue proposes for framework dependent apps - with the SDK printing out a message that the app will expect the runtime (and ideally a version + link) at a relative location `some/path`. It's not great, as the app doesn't run, but then again - you asked for it :-). For this reason though I would only support this for `dotnet publish`. I would try to avoid this for `dotnet build` as that would effectively mean that the app can't be debugged using the normal means (like `F5` in VS).

For self-contained apps, the right thing to do would be for the SDK to put the "runtime" into that subfolder. The interesting question then is "what is a runtime?". Is it just the runtime packs and it would leave NuGets and project references in the app root, or does it mean everything "code-like?". This one is probably simpler with regard to end-to-end behavior as the produced app would just work.

Under the hood these two features would probably use the same apphost customization logic, which is mostly what this issue is proposing. But in a way that could be considered an implementation detail.

**Comment by Symbai** [comment]
> The interesting question then is "what is a runtime?". Is it just the runtime packs and it would leave NuGets and project references in the app root, or does it mean everything "code-like?"

That one is pretty easy to answer: Let the developers decide. Just like you did with self-contained to include native files for example. See my comment https://github.com/dotnet/sdk/issues/10366#issuecomment-998533665 by default it copies the whole runtime files. Per opt-in it also copies all .NET for example nuget packages. Per another opt-in it also copies all native files.

I don't know how the whole "publish" process works. But to me it doesn't seem to be a problem by just copying the RUNTIME files into a different folder. Then add a JSON property the `.runtime.json` file specifying the new relative location for the apphost to look at.  Theoretically its just the apphost executable that needs to know where the runtime files are. And for self-contained it needs the "publish process" to copy the runtime files to this folder. That should be it. For example the apphost patcher mentioned in https://github.com/dotnet/runtime/issues/3453 does exactly this and it still works. But instead of hard coding the file into the apphost executable it could just be a property in the JSON file.

**Comment by Symbai** [comment]
> Personally I would want to avoid putting this information into a JSON file. Parsing JSON is a non-trivial performance hit and this is about app startup which is super perf sensitive.

You have to be honest that adding ONE additional JSON property isn't a big deal. When it comes to JSON performance I would rather be worried about the already added runtime assembly files information. Such as:

          "Microsoft.VisualBasic.Core.dll": {
            "assemblyVersion": "11.0.0.0",
            "fileVersion": "6.0.121.56705"
          },

For a WPF application these alone makes `>1,000` lines. If this has been added, then +1 line wouldn't cause any noticeably impact. However this was just my suggestion about how this can be done without too much effort. You can also choose another way. Just keep in mind that we have now 3 (with this one 4) requests for almost the same feature since .NET (Core) 3. Looking for the perfect solution which delays this feature until .NET v20 isn't ideal. Maybe doing small steps could help.

**Comment by vitek-karas** [comment]
I agree that adding one more property to already existing JSON is not a big deal, but as I explained above this actually means adding:
* JSON parser to apphost (bigger binary - which ships with every .NET app in the world... so this would be a hard sell, there are already people complaining that it's 150KB which doesn't really do anything)
* Search/open a new file - this shows as a noticable perf thing especially on Windows
* Parse it as JSON - this is probably the least important part - should be really fast

I'm honestly not trying to delay anything here. At the same time, we've been bitten many times by "Features" which were halfway done. So I'm trying to figure out what solving the whole end-to-end problem means in this case. It doesn't necessarily mean we have to implement all of it in one release, but I would rather have a rough idea how it should look overall.

**Comment by Symbai** [comment]
> * JSON parser to apphost

Ok, this was a misunderstanding then. I thought the apphost would already parse the JSON files. If that's not the case then I will withdraw my suggestion and agree with you that it doesn't make sense.

**Comment by vitek-karas** [comment]
My apologies - I should have made it clearer in my initial reply.

**Comment by agocke** [comment]
Yeah this sounds quite a lot like what @vitek-karas and I came up with independently. The coincidence is probably a positive sign that the simple version is workable. I agree that for adding more SDK support it would be good to get a sense of how the "user friendly" version of this feature would work.

**Comment by richlander** [comment]
@vitek-karas -- how is adding `DOTNET_ROOT` to `runtimeconfig.json` more expensive than the existing GC mode, or for that matter the framework version. It would assume that this value would be similar. Ideally, all ENVs that the host honors would be specified in `runtimeconfig.json`. Isn't that reasonable?

**Comment by vitek-karas** [comment]
@richlander It's a componentization issue. Currently the system works like this:
* apphost does only one thing - determine the install root (using registry and so on), note that there's no file parsing involved in this. It then loads `hostfxr` from that location and calls into it to do the trest.
* `hostfxr` has all the logic to parse `.runtimeconfig.json` and so on.

So in order to support install location to be specified in the `.runtimeconfig.json` we would have to either:
* Add JSON parsing code into apphost and parse it there (and then again in hostfxr for the second time unless we move more code into apphost) - this is problematic because it will increase the size of apphost (which is already 150KB which is quite large). Not counting the perf hit of reading the file twice and so on. Note that we would have to do this for all applications.
* In theory it would be possible for hostfxr to "change" the install root - we kind of already do this due to multi-level lookup. The problem with this approach is the forward/backward compatibility issue - which exists today, but this would make it more frequent. And another problem is if there's no global install, then the apphost may not be able to find any `hostfxr` on its own - so it would need to read the `.runtimeconfig.json` anyway.
* Last option is to have two apphosts - developer would be required to "Enable" this feature during build and SDK would use different apphost which would have the code to parse the JSON in it. This would introduce more complexity for us - we currently have 2 apphosts (the normal one and then the single-file one), we would have to add another (across all platforms/architectures).

It's definitely doable, but it's quite a bit more complex to implement compared to basically any other new property added to `.runtimeconfig.json`.

**Comment by Symbai** [comment]
How exactly is the app host executable generated? Is it a fixed executable that it taken from somewhere and copied to output folder? Or this is build dynamically for each project? If build dynamically, then why cannot the custom path be patched/written to the executable on build?

For example the apphost patcher linked in one of the other requests, which again still works!, just patches the hardcoded path in the apphost executable. The benefit is:

- Still no JSON parsing. Literally same app host size (+ few bytes for longer path string)
- Still one apphost
- No changes to hostfxr

While this solves the request for moving ALL files except the app host into a sub folder. It probably won't do much more than that, unless you add more stuff to the apphost which can be "patched" on build.

**Comment by vitek-karas** [comment]
It's a prebuilt executable (coming from NuGet typically) which is "patched" during the build. I absolutely agree that patching a hardcoded relative path to the apphost is the cheapest option right now. It's a relatively small change to apphost and then a small change to the "patcher". This is BTW exactly what the issue suggests above.

And it has exactly the benefits you mention.

Just to clarify: We currently don't build the apphost during app build, because the only sane way of doing that would require we use platform linker to produce the final executable. That has its own advantages, but it has one major disadvantage: The SDK would have a dependency on platform linker, which is currently not the case (at least for the default workloads). So far we've been trying to avoid adding such dependency.

**Comment by Symbai** [comment]
> And it has exactly the benefits you mention.

When I mentioned it for the first time you said in another [post](https://github.com/dotnet/runtime/issues/3453#issuecomment-509249471) that it comes with the tag of "can be broken at anytime". I can also see that you created a reference issue requesting the customization of the path. Which then got closed for a reference of another tracking issue which has no link, I assume its in a private repo or something?

Whats the status of this customization and would you believe that this could be a solution you would consider to add? It would be amazing if this could be done by adding a single line in `csproj` file for developers and we longer have to rely on third party tools with a chance of it stops working in any update.

**Comment by vitek-karas** [comment]
My comments above are related to the situation where this is implemented as a feature - in which case the necessary customization would be wired through the SDK and become an official supported functionality (unlike the existing way referred to by me before).

As noted above I think we need to figure out the desired end-to-end behavior first. I do want to avoid situation where we add some supported feature now, and then change it in the next release because it doesn't fit the end-to-end. It might still happen, but hopefully we can reduce the risk of that now.

Adding the one small customization in itself doesn't solve the problem for most people, only for the select few who are willing to add several addition custom build steps to copy files around and wire everything up correctly.

Also of note: There are two different customization we're discussing here:
* Ability to customize the install location to look for runtimes - basically an equivalent of built-in `DOTNET_ROOT` - typically would be used by framework-dependent apps.
* Ability to customize the location of everything but the apphost (which is what you refer to in your previous comment) - typically would be used for self-contained apps. This is not required to fulfill the ask of this issue (at least the way I understand it).

**Comment by vuplea** [comment]
We've also had to deploy multiple AppHosts in self-contained mode (we build a user-mode MSI installer) that have to reuse the framework files. We resorted to building all the apps in the same folder, which is quite cumbersome (through a custom solution, we ensure that all common `ReferenceCopyLocalPaths` between exe projects match).

And another problem for which a sharable runtime directory would help is described in https://github.com/dotnet/runtime/issues/64538.

**Comment by agocke** [comment]
We're still strongly interested in this, but haven't had time to explore the solution space as Vitek mentioned

**Comment by richlander** [comment]
I filed an issue to track the broader topic for .NET 8 -> https://github.com/dotnet/runtime/issues/71282

**Comment by SamJongenelen** [comment]
Another example would be running a ServiceFabric cluster containing a lot of dotnet modules, all using the exact same runtime. 
My interest would purely be reducing the total applications` size.

**Comment by richlander** [comment]
Can you do some back-of-the-napkin math and tell us how much size you think you'd save?

**Comment by SamJongenelen** [comment]
~50mb per module, in my case 20 modules so 1000mb. Roughly ofcourse.

**Comment by lindexi** [comment]
My software uses the mulitple process mode which need some Exe entry file. Per EXE entry file is one of the entry assembly. The bad news is today we can not find any practices about multiple entry assembly in dotnet. The core problem is we will reference difference version assembly which will be conflicting. To solve it, we have to publish each entry project independently.

Without share the dotnet runtime folder, we will create lots of duplicate files which is the part of dotnet runtime.

My software uses 20 entry assembly, that means we will create 20 duplicate dotnet runtime files, about `123MB x 20 = 2460MB`

**Comment by richlander** [comment]
> My software uses the multiple process mode which need some Exe entry file.

Could you elaborate on that?

**Comment by lindexi** [comment]
@richlander The main Exe entry files of my application software are:

- TheMainApp.exe : The main entry file.
- X86InteractiveProcess.exe: Using IPC to run the x86 native libraries.
- Update.exe : To help the app to update.

- Business1.exe : The business 1 entry file;
- Business2.exe : The business 2 entry file;
- ...

**Comment by Symbai** [comment]
> This does not include an SDK experience (you need to create an [install layout](https://github.com/dotnet/designs/blob/main/accepted/2020/install-locations.md#net-core-install-layout) yourself)

Is there a convient way to automate this so when I run `dotnet publish` the runtime files are being placed into a sub folder? Because if I have to manually move files every time I publish I'm definitely not going to use that. Especially with (unsupported) third party libraries and changing runtime files this easily leads to mistakes.

**Comment by lindexi** [comment]
Thank you @elinor-fung

**Comment by Symbai** [comment]
Is this feature at least planned or tracked somewhere? I'd really like to specify a relative path in my `csproj` and on build it moves the runtime and libraries into this sub folder so I get a clean folder structure for end users.

**Comment by elinor-fung** [comment]
I took this issue as just the dotnet root ask (basically specifying `DOTNET_ROOT` as you would for framework-dependent apps, but at build-time).

For the SDK layout customization, in dotnet/sdk:
- https://github.com/dotnet/sdk/issues/10366

The shared self-contained deployment is largely an SDK artifact generation and layout concern, so I think of that issue as the bulk of the work.

For tracking in the runtime repo, we have:
- https://github.com/dotnet/runtime/issues/3453 (one you filed a long time ago)

(There is also https://github.com/dotnet/runtime/issues/53834, which I think is is a superset of the above)

I don't think we have this actually planned - cc @richlander @baronfel in case they are aware of planning that I am not.

**Comment by Symbai** [comment]
I already checked the two issues you also posted earlier but both doesn't really seem to cover this case. What I ask for would probably be something to add to the SDK. Its about automatically moving the runtime into a sub folder, based on the recently added apphost functionality, if perhaps a new flag is set. I also don't see this planned or tracked either that's why I was asking. I could create a new issue for that.

**Comment by agocke** [comment]
A new issue means sense to me if your scenario is not mentioned.
